Source/Particles/Collision/BinaryCollision/BinaryCollision.H

```diff
@@ -335,31 +359,98 @@ public:
                     index_type const cell_stop_1  = cell_offsets_1[i_cell+1];
                     index_type const cell_half_1 = (cell_start_1+cell_stop_1)/2;
 
-                    // Same but for the pairs
-                    index_type const cell_start_pair = have_product_species?
-                                                           p_pair_offsets[i_cell] : 0;
-
                     // Do not collide if there is only one particle in the cell
                     if ( cell_stop_1 - cell_start_1 <= 1 ) { return; }
 
                     // shuffle
                     ShuffleFisherYates(
                         indices_1, cell_start_1, cell_half_1, engine );
+                }
+            );
+
+            // Loop over cells
+            amrex::ParallelForRNG( n_independent_pairs,
+                [=] AMREX_GPU_DEVICE (int i_coll, amrex::RandomEngine const& engine) noexcept
+                {
+                    // to avoid type mismatch errors
+                    auto ui_coll = (index_type)i_coll;
+
+                    // compute i_cell = upper_bound(p_coll_offsets[0:n_independent_
+                    int i_cell = 0;
+                    {
+                        int left = 0;
+                        int right = n_cells;
+
+                        for (;;)
```

Comment:
Yes of course, if `amrex` implements the same binary search logic, it can definitely be replaced. Yes, it can be replaced with a `while(left < right)` loop or the same condition could be added here as well.